docs/application-tutorial/game_of_life.py

```diff
@@ -0,0 +1,551 @@
+import sys
+import time
+import math
+import numpy as np
+
+# Hide pygame prompt
+from os import environ
+
+environ["PYGAME_HIDE_SUPPORT_PROMPT"] = "1"
+
+import pygame
+import argparse
+
+from concrete import fhe
+
+# Constants of the game of life
+weights_method_3b_a = np.array(
+    [
+        [1, 1, 1],
+        [1, 0, 1],
+        [1, 1, 0],
+    ]
+)
+weights_method_3b_b = np.array(
+    [
+        [0, 0, 0],
+        [0, 0, 0],
+        [0, 0, 1],
+    ]
+)
+table_next_cell_3b_a = [i if i <= 3 else 4 for i in range(8)]
+table_next_cell_3b_b = [i - 1 if i in [2, 3] else 0 for i in range(6)]
+table_next_cell_3b_c = [int(i in [2, 3]) for i in range(4)]
+
+weights_method_4b = np.array(
+    [
+        [1, 1, 1],
+        [1, 0, 1],
+        [1, 1, 1],
+    ]
+)
+table_next_cell_4b_a = [i - 1 if i in [2, 3] else 0 for i in range(9)]
+table_next_cell_4b_b = [int(i in [2, 3]) for i in range(4)]
+
+weights_method_5b = np.array(
+    [
+        [1, 1, 1],
+        [1, 9, 1],
+        [1, 1, 1],
+    ]
+)
+table_next_cell_5b = [int(i in [3, 9 + 2, 9 + 3]) for i in range(18)]
+
+weights_method_basic = np.array(
+    [
+        [1, 1, 1],
+        [1, 0, 1],
+        [1, 1, 1],
+    ]
+)
+table_next_cell_basic_a = [int(i in [3]) for i in range(9)]
+table_next_cell_basic_b = [int(i in [2, 3]) for i in range(9)]
+
+
+# CP tables
+table_cp_next_cell_3b_a = fhe.LookupTable(table_next_cell_3b_a)
+table_cp_next_cell_3b_b = fhe.LookupTable(table_next_cell_3b_b)
+table_cp_next_cell_3b_c = fhe.LookupTable(table_next_cell_3b_c)
+
+table_cp_next_cell_4b_a = fhe.LookupTable(table_next_cell_4b_a)
+table_cp_next_cell_4b_b = fhe.LookupTable(table_next_cell_4b_b)
+
+table_cp_next_cell_5b = fhe.LookupTable(table_next_cell_5b)
+
+table_cp_next_cell_basic_a = fhe.LookupTable(table_next_cell_basic_a)
+table_cp_next_cell_basic_b = fhe.LookupTable(table_next_cell_basic_b)
+
+# Function to workaround the miss of padding in CP
+def by_hand_padding(original_grid, res):
+    padded_res = fhe.zeros(original_grid.shape)
+
+    original_grid_shape = original_grid.shape
+    assert padded_res.shape[0:2] == (1, 1)
+    padded_res[0, 0, 1 : original_grid_shape[2] - 1, 1 : original_grid_shape[3] - 1] = res
+
+    assert original_grid.shape == padded_res.shape
+
+    return padded_res
+
+
+# Function for Game of Life
+@fhe.compiler({"grid": "encrypted"})
+def update_grid_method_3b(grid):
+
+    # Method which uses two first TLU of 3 bits and a third TLU of 2 bits
+
+    # This is to workaround the fact that we have no pad option in fhe.conv
+    do_padded_fix = True
+
+    # Compute the sum of 7 elements
+    convoluted_grid = fhe.conv(
+        grid,
+        weights_method_3b_a.reshape(1, 1, *weights_method_3b_a.shape),
+        strides=(1, 1),
+        dilations=(1, 1),
+        group=1,
+        pads=(1, 1, 1, 1) if not do_padded_fix else (0, 0, 0, 0),
+    )
+
+    if do_padded_fix:
```

Comment:
define a your conv function. (so you don't need to repeat the workaround and you can fix the strides and dilations groups as there are constants)

Reply:
ok done